### PR TITLE
Backend auth

### PR DESCRIPTION
**Pull request by JDziewonski98**
# Summary

Backend firebase token authentication to only allow logged in (admin) users to use the modify data endpoints.
When you refresh the front end, you stay logged in but the token gets wiped, not sure how to fix.
Also the token is only valid for 1h, can there be a scheduled function on frontend to refresh it?
Don't merge this PR, it needs a firebase key file to function, and it is best practice to not push it, need to figure out how to get it on the deployment securely.
I tried a few different designs for the auth on backend, and found a middleware pattern that is cleanest and doesn't break tests.
On frontend I made the context instances type 'any' otherwise it didn't let me grab the token?
## Test Plan
Manually

## Related Issues
Backend Auth

### CODE REVIEW

File: backend/src/auth/AuthMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * Middleware that can be added at route level to add firebase authentication to a controller
+ * Expects a valid firebase jwt token in headers.authorization
+ * @param req HTTP request
+ * @param res HTTP response
+ * @param next reference to controller function to be executed if auth passes
+ */
+export const authMiddleware = async (req: Request, res: Response, next: any) => {
+  try {if (req.headers.authorization) {
+    await FBAuth.auth()
+      .verifyIdToken(req.headers.authorization)
```
Comment (tedspare) [inline backend/src/auth/AuthMiddleware.ts:15]: nice

File: frontend/src/App/App.tsx [hunk 2]
```diff
@@ -130,7 +133,7 @@ function App() {
 
   return (
     <div className="App">
-      <Context.Provider value={{ user: user }}>
+      <Context.Provider value={{ user: user, token : JWTtoken }}>
```
Comment (tedspare) [inline frontend/src/App/App.tsx:136]: nice

File: frontend/src/TourForm/TourForm.tsx [hunk 3]
```diff
@@ -91,7 +90,7 @@ function TourForm({ tour, muralsData, handleCancel }: ITourFormProps) {
   /**
    * Enable editable form fields for admin users
    */
-  const userContext = useContext(Context)
+  const userContext: any = useContext(Context)
   useEffect(() => setIsAdmin(!!(userContext as any).user), [userContext]);
```
Comment (tedspare) [inline frontend/src/TourForm/TourForm.tsx:94]: were the `:any`s added intentionally? if so they should simplify the next line to just `... !!userContext.user ...`

**Review by tedspare — COMMENTED** [review]
generally non-invasive solution, kudos

LGTM once we figure out how to persist and find files on the EC2 instance

File: frontend/src/TourForm/TourForm.tsx (hunk 3 above)
Comment (JDziewonski98) [reply frontend/src/TourForm/TourForm.tsx:94]: If I don't add the any I get an error on userContext.token.i

File: .ebextensions/privatekey.config [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+              OptionName: "IamInstanceProfile"
+              DefaultValue: "aws-elasticbeanstalk-ec2-role"
+files:
+  # Private key
+  "/mu-auth-73f4d05ed60d.json":
+    mode: "000400"
+    owner: root
+    group: root
+    authentication: "S3Auth"
+    source: https://elasticbeanstalk-us-east-1-115630815585.s3.amazonaws.com/privatekey/mu-auth-73f4d05ed60d.json
```
Comment (tedspare) [inline .ebextensions/privatekey.config:20]: should this url be a string?

**Review by tedspare — APPROVED** [review]
LGTM. Is this what peak dev performance looks like?

File: .ebextensions/privatekey.config (hunk 4 above)
Comment (JDziewonski98) [reply .ebextensions/privatekey.config:20]: https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/https-storingprivatekeys.html Not per the tutorial

**Pull request merged**